File: pi/extensions/tasks/insert.ts
/**
 * Cross-extension task insertion helper.
 *
 * `buildTaskBlock` is the single source of truth for assembling an
 * org-mode task block from structured fields. Consumed today by the
 * Jira `jira_clone_apply` tool and exposed publicly so future
 * cross-tracker integrations (github / linear / gitlab / `/jira create`)
 * never reimplement priority mapping, drawer ordering, or label
 * tagging.
 *
 * `buildTaskBlock` remains pure for compatibility tests and future
 * tracker integrations. The file-side `insertTaskIntoFile` entrypoint
 * is now a thin compatibility shim over `ot create`, so durable
 * insertion/idempotency behaviour lives in the Babashka CLI.
 */

import { randomUUID } from "node:crypto";
import { basename, isAbsolute, resolve } from "node:path";
import {
  createdLogEntry,
  formatOrgTimestamp,
} from "./parser.ts";
import { runOt, type OtEnvelope } from "./ot.ts";

/** Args accepted by {@link buildTaskBlock}. */
export interface BuildTaskArgs {
  /** Task heading text. Required. Trailing whitespace is trimmed. */
  summary: string;
  /**
   * Priority *name* (Jira convention). One of:
   * `Highest` → `#A`, `High` → `#B`, `Medium` → `#C`,
   * `Low`|`Lowest` → `#D`. Anything else (including null/undefined or
   * the empty string) yields no priority cookie.
   *
   * Case-insensitive. Whitespace-trimmed.
   */
  priorityName?: string | null;
  /** Body text inserted verbatim after the drawer. May be empty/null. */
  body?: string | null;
  /**
   * External-tracker references to write into `:LINKED_ISSUES:`. Tokens
   * are written verbatim, whitespace-joined. Empty array → no
   * `:LINKED_ISSUES:` line.
   */
  linkedIssues?: string[] | null;
  /**
   * Labels appended to the heading as org tags
   * (`:label1:label2:` after the summary). Tokens are emitted verbatim;
   * callers are responsible for sanitising disallowed characters
   * (org tags accept `[a-zA-Z0-9_@]`).
   * Empty array → no tags.
   */
  labels?: string[] | null;
  /**
   * When set, the assembled block is rendered as a level-3 heading
   * (i.e. a subtask under an existing parent task). When unset, the
   * block is level-2 (a top-level task under a section heading).
   *
   * The parent ID itself is *not* embedded in the drawer — it's used
   * solely to pick the heading level. The file-side helper is
   * responsible for placing the block inside the parent task's
   * subtree.
   */
  parentId?: string | null;
  /**
   * Override the generated `:CUSTOM_ID:` UUID. Used by the file-side helper
   * to surface the new task's ID back to its caller, and by tests to
   * keep snapshots deterministic. Defaults to a fresh UUID v4.
   */
  id?: string;
  /**
   * Override the `:CREATED:` timestamp body (without surrounding
   * brackets). Defaults to the current local time formatted as
   * `YYYY-MM-DD Day HH:MM`. Injectable for tests.
   */
  createdAt?: string;
}

/** Output of {@link buildTaskBlock}. */
export interface BuiltTaskBlock {
  /** Heading line, e.g. `** TODO [#A] Summary :foo:bar:`. */
  heading: string;
  /**
   * Properties drawer block, including the `:PROPERTIES:` / `:END:`
   * fences. Always emitted (the `:CUSTOM_ID:` line guarantees non-empty
   * content). Multi-line string, no trailing newline.
   */
  drawer: string;
  /**
   * Body text as supplied by the caller, normalised to have no
   * leading or trailing newlines. May be the empty string.
   */
  body: string;
  /**
   * Fully-assembled org block ready to splice into a section. Always
   * ends with a single trailing newline so the file-side helper can
   * concatenate without bookkeeping. Layout:
   *
   * ```
   * ** TODO [#A] Summary :foo:bar:
   * :PROPERTIES:
   * :CUSTOM_ID: <uuid>
   * :CREATED: [<timestamp>]
   * :LINKED_ISSUES: KEY1 KEY2
   * :END:
   * <body>
   * ```
   *
   * Body is omitted when empty; the trailing newline still applies.
   */
  block: string;
  /** The `:CUSTOM_ID:` UUID written into the drawer. */
  id: string;
}

/**
 * Map a Jira priority name to an org priority cookie character
 * (`A`/`B`/`C`/`D`), or null when the input doesn't match a known bucket.
 *
 * Exported for parity-checking in tests / future tracker integrations.
 */
export function mapPriorityName(name: string | null | undefined): string | null {
  if (!name) return null;
  const normalised = name.trim().toLowerCase();
  switch (normalised) {
    case "highest": return "A";
    case "high":    return "B";
    case "medium":  return "C";
    case "low":
    case "lowest":  return "D";
    default:        return null;
  }
}

/**
 * Render a list of labels as an org tag suffix (`:foo:bar:`), or the
 * empty string when there are no labels.
 */
function renderTagSuffix(labels: string[] | null | undefined): string {
  if (!labels || labels.length === 0) return "";
  const filtered = labels.filter((l) => l && l.length > 0);
  if (filtered.length === 0) return "";
  return `:${filtered.join(":")}:`;
}

/**
 * Build an org-task block (heading + drawer + body) from structured
 * fields. Pure: no file I/O, no random side-effects beyond the default
 * UUID + timestamp.
 *
 * The status is hard-coded to `TODO`. Cloned issues always start in
 * the local TODO state regardless of their tracker-side status — local
 * status is a *contributor's* signal, not the tracker's.
 */
export function buildTaskBlock(args: BuildTaskArgs): BuiltTaskBlock {
  const summary = args.summary.trimEnd();
  if (!summary) {
    throw new Error("buildTaskBlock: summary is required");
  }

  const id = args.id ?? randomUUID();
  const createdAt = args.createdAt ?? formatOrgTimestamp();
  const level = args.parentId ? 3 : 2;
  const stars = "*".repeat(level);

  const priorityChar = mapPriorityName(args.priorityName);
  const priorityCookie = priorityChar ? `[#${priorityChar}] ` : "";
  const tagSuffix = renderTagSuffix(args.labels);
  const heading =
    `${stars} TODO ${priorityCookie}${summary}` +
    (tagSuffix ? ` ${tagSuffix}` : "");

  const drawerLines = [
    ":PROPERTIES:",
    `:CUSTOM_ID: ${id}`,
    `:CREATED: [${createdAt}]`,
  ];
  const linkedIssues = (args.linkedIssues ?? []).filter((t) => t && t.length > 0);
  if (linkedIssues.length > 0) {
    drawerLines.push(`:LINKED_ISSUES: ${linkedIssues.join(" ")}`);
  }
  drawerLines.push(":END:");
  const drawer = drawerLines.join("\n");

  const body = (args.body ?? "").replace(/^\n+/, "").replace(/\n+$/, "");

  const blockLines = [heading, drawer, ":LOGBOOK:", createdLogEntry(createdAt), ":END:"];
  if (body.length > 0) blockLines.push(body);
  const block = blockLines.join("\n") + "\n";

  return { heading, drawer, body, block, id };
}

// ─── File-side insertion + idempotency ──────────────────────────────
//
// `insertTaskIntoFile` is the cross-extension entry point consumed by
// `jira_clone_apply` (today) and any future tracker integration. Kept
// here (rather than in `index.ts`) so:
//
// 1. It has no `pi-tui` / `pi-coding-agent` dependency and can be
//    unit-tested directly via `tsx`.
// 2. The cross-extension contract is reusable as a plain JS function
//    without round-tripping through the LLM tool registry.

/** Recognised duplicate / placement failure modes. */
export type InsertErrorReason =
  | "duplicate"
  | "section_not_found"
  | "empty_summary"
  | "file_unreadable"
  | "path_outside_project";

/** Args accepted by {@link insertTaskIntoFile}. */
export interface InsertTaskArgs extends BuildTaskArgs {
  /**
   * Absolute path of the org file to insert into. Most callers use
   * `<cwd>/TASKS.org`; the field is left flexible so the same helper
   * can splice into a `TASKS.local.org` draft or a linked plan.
   */
  file: string;
  /**
   * Section heading text under which the task is appended
   * (e.g. `"Improvements"`). Matched as a level-1 heading in the
   * target file. Tags on the heading line are tolerated.
   */
  section: string;
  /**
   * When true and `section` does not yet exist in the file, append a
   * new `* <section>` heading at the end of the file before splicing.
   * Default: false.
   */
  allowCreateSection?: boolean;
  /**
   * Additional org files to scan for `:LINKED_ISSUES:` collisions.
   * The target file is always scanned. Callers typically pass the
   * sibling file (e.g. `TASKS.local.org` when inserting into
   * `TASKS.org`, and vice-versa) so duplicates are detected
   * regardless of which slot the previous clone landed in.
   *
   * Imports referenced via `#+IMPORT:` from any scanned file are
   * recursively walked.
   */
  alsoScan?: string[];
  /** Project root used to sandbox target and scan paths. Defaults to cwd. */
  projectRoot?: string;
}

/** Successful insertion result. */
export interface InsertSuccess {
  status: "inserted";
  /** UUID written into the new task's `:CUSTOM_ID:` drawer line. */
  id: string;
  /** Absolute path of the file mutated. */
  file: string;
  /** 1-indexed line where the new heading lives after insertion. */
  line: number;
}

/** Refusal — duplicate :LINKED_ISSUES: token already present. */
export interface InsertDuplicate {
  status: "duplicate";
  /** `:CUSTOM_ID:` of the pre-existing task that owns the conflicting token. */
  existingId: string | null;
  /** Absolute path of the file containing the pre-existing task. */
  existingFile: string;
  /** The `:LINKED_ISSUES:` token that triggered the refusal. */
  conflictingToken: string;
}

/** Refusal — section heading not found and `allowCreateSection` false. */
export interface InsertSectionMissing {
  status: "section_not_found";
  file: string;
  section: string;
}

/** Refusal — caller mis-configured the request. */
export interface InsertError {
  status: "error";
  reason: InsertErrorReason;
  message: string;
}

export type InsertResult =
  | InsertSuccess
  | InsertDuplicate
  | InsertSectionMissing
  | InsertError;

/**
 * Insert a new task into an org file under the named section,
 * refusing on duplicate `:LINKED_ISSUES:` overlap.
 *
 * Pure-ish: reads `args.file` (and `args.alsoScan`) for duplicate
 * detection, writes back `args.file` on success. No UI.
 */
export async function insertTaskIntoFile(
  args: InsertTaskArgs,
): Promise<InsertResult> {
  if (!args.summary || args.summary.trim().length === 0) {
    return {
      status: "error",
      reason: "empty_summary",
      message: "`summary` is required and must be non-empty.",
    };
  }

  const projectRoot = isAbsolute(args.projectRoot ?? "")
    ? args.projectRoot!
    : resolve(args.projectRoot ?? ".");

  const cmd = ["create", args.summary];
  if (args.section) cmd.push("--section", args.section);
  if (args.priorityName) cmd.push("--priority", args.priorityName);
  if (args.body) cmd.push("--body", args.body);
  if (args.parentId) cmd.push("--parent", args.parentId);
  if (args.allowCreateSection) cmd.push("--allow-create-section");
  if (args.id) cmd.push("--id", args.id);
  if (args.createdAt) cmd.push("--created-at", args.createdAt);
  for (const token of args.linkedIssues ?? []) {
    if (token) cmd.push("--linked-issue", token);
  }
  for (const label of args.labels ?? []) {
    if (label) cmd.push("--tag", label);
  }
  for (const scanPath of args.alsoScan ?? []) {
    if (scanPath) cmd.push("--also-scan", scanPath);
  }

  // `ot create --local` writes to <root>/TASKS.local.org; otherwise
  // pass --tasks <file> as a global override so the legacy JS entry
  // point remains file-parameter compatible for TASKS.org and tests.
  const globalArgs = basename(args.file) === "TASKS.local.org"
    ? []
    : ["--tasks", args.file];
  if (basename(args.file) === "TASKS.local.org") cmd.push("--local");

  let env: OtEnvelope<{ id: string; file: string; line: number }>;
  try {
    env = await runOt(cmd, { root: projectRoot, globalArgs });
  } catch (err) {
    return {
      status: "error",
      reason: "file_unreadable",
      message: (err as Error).message,
    };
  }

  if (env.ok) {
    return {
      status: "inserted",
      id: env.result.id,
      file: env.result.file,
      line: env.result.line,
    };
  }

  const { code, message, file, details } = env.error;
  switch (code) {
    case "duplicate-linked-issue":
      return {
        status: "duplicate",
        existingId: (details?.existingId as string | null | undefined) ?? null,
        existingFile: (details?.existingFile as string | undefined) ?? file ?? args.file,
        conflictingToken: (details?.conflictingToken as string | undefined) ?? "",
      };
    case "section-not-found":
      return {
        status: "section_not_found",
        file: file ?? args.file,
        section: (details?.section as string | undefined) ?? args.section,
      };
    case "empty-summary":
      return { status: "error", reason: "empty_summary", message };
    case "path-outside-project":
      return { status: "error", reason: "path_outside_project", message };
    case "unknown-task":
      return { status: "error", reason: "file_unreadable", message };
    default:
      return { status: "error", reason: "file_unreadable", message };
  }
}

File: skills/org-tasks/scripts/src/org_tasks/insert.clj
(ns org-tasks.insert
  "Task insertion + idempotency across `:LINKED_ISSUES:` tokens.

  Pure where possible: `build-task-block` is fully deterministic given
  `id` + `created-at` overrides; `insert-task-into-file` reads `target`
  (+ optional `also-scan`), checks for duplicates, splices the new
  block under the requested section, and writes the result atomically.

  Public surface:

    map-priority-name        :: Jira name -> 'A'|'B'|'C'|'D' or nil
    build-task-block         :: args -> {:heading, :drawer, :body,
                                         :block, :id}
    insert-task-into-file    :: args -> result map (see :status keys)"
  (:require [babashka.fs :as fs]
            [clojure.string :as str]
            [org-tasks.effective :as effective]
            [org-tasks.links :as links]
            [org-tasks.loader :as loader]
            [org-tasks.parser :as parser]
            [org-tasks.paths :as paths]
            [org-tasks.tree :as tree]))

;; ── Priority mapping ───────────────────────────────────────────────

(defn map-priority-name
  "Map a Jira priority name (or short form) to an org cookie character.

  Highest → A, High → B, Medium → C, Low|Lowest → D. Single-letter
  inputs `A`/`B`/`C`/`D` (case-insensitive) pass through. Anything
  else yields nil."
  [name]
  (when (and name (seq name))
    (let [n (str/lower-case (str/trim name))]
      (case n
        ("a" "highest") "A"
        ("b" "high")    "B"
        ("c" "medium")  "C"
        ("d" "low" "lowest") "D"
        nil))))

;; ── Heading + drawer assembly ──────────────────────────────────────

(defn- render-tag-suffix [labels]
  (let [filtered (filterv #(and % (seq %)) (or labels []))]
    (if (empty? filtered)
      ""
      (str ":" (str/join ":" filtered) ":"))))

(defn build-task-block
  "Build an org-task block from structured fields. Status is always
  TODO; cloned/imported tasks always start in local TODO regardless
  of the source tracker.

  Args:
    :summary       - heading text, required.
    :priority-name - Jira priority name; passed through map-priority-name.
    :body          - optional body text, normalised to no leading/trailing
                     newlines.
    :linked-issues - sequence of `:LINKED_ISSUES:` org-link tokens.
    :labels        - sequence of org tag strings.
    :parent-id     - when set, render the block as a level-3 heading.
    :id            - override the generated `:CUSTOM_ID:` (test injection).
    :created-at    - override the `:CREATED:` timestamp body (test injection).

  Returns `{:heading, :drawer, :body, :block, :id}`. `:block` ends in
  exactly one trailing newline."
  [{:keys [summary priority-name body linked-issues labels parent-id id created-at]}]
  (let [summary (str/trimr (or summary ""))]
    (when (empty? summary)
      (throw (ex-info "build-task-block: summary is required" {})))
    (let [id          (or id (str (random-uuid)))
          created-at  (or created-at (parser/format-org-timestamp))
          level       (if parent-id 3 2)
          stars       (apply str (repeat level "*"))
          priority-char (map-priority-name priority-name)
          priority-cookie (if priority-char (str "[#" priority-char "] ") "")
          tag-suffix  (render-tag-suffix labels)
          heading     (str stars " TODO " priority-cookie summary
                           (when (seq tag-suffix) (str " " tag-suffix)))
          base-drawer [":PROPERTIES:"
                       (str ":CUSTOM_ID: " id)
                       (str ":CREATED: [" created-at "]")]
          linked      (filterv #(and % (seq %)) (or linked-issues []))
          drawer-lines (cond-> base-drawer
                         (seq linked) (conj (str ":LINKED_ISSUES: "
                                                 (str/join " " linked))))
          drawer      (str/join "\n" (conj drawer-lines ":END:"))
          body        (-> (or body "")
                          (str/replace #"^\n+" "")
                          (str/replace #"\n+$" ""))
          block-lines (cond-> [heading drawer ":LOGBOOK:"
                               (parser/created-log-entry created-at)
                               ":END:"]
                        (seq body) (conj body))
          block       (str (str/join "\n" block-lines) "\n")]
      {:heading heading
       :drawer  drawer
       :body    body
       :block   block
       :id      id})))

;; ── Section splice ─────────────────────────────────────────────────

(defn- section-heading? [^String line ^String section]
  (when-let [m (re-matches #"^\*\s+(.+?)\s*$" line)]
    (let [text (m 1)
          ;; Strip trailing `:tag1:tag2:`
          stripped (if-let [tm (re-find #"\s+:[\w@:]+:\s*$" text)]
                     (str/trimr (subs text 0 (- (count text) (count tm))))
                     (str/trimr text))]
      (= (str/trim stripped) (str/trim section)))))

(defn- next-top-level-heading-idx [lines ^long from]
  (some (fn [i]
          (when (re-matches #"^\*\s+\S.*$" (nth lines i))
            i))
        (range from (count lines))))

(defn- splice-into-section
  "Splice `block` into `content` under `* <section>`. Returns
  `{:content, :line}` (1-indexed line of the new heading) or nil
  when the section is absent and `allow-create-section?` is false."
  [^String content ^String section ^String block ^Boolean allow-create-section?]
  (let [lines (vec (str/split content #"\n" -1))
        heading-idx (some (fn [i]
                            (when (section-heading? (nth lines i) section)
                              i))
                          (range (count lines)))
        block-lines (vec (str/split (str/replace block #"\n+$" "") #"\n"))]
    (cond
      ;; New section append (when allowed)
      (and (nil? heading-idx) allow-create-section?)
      (let [trimmed (loop [v lines]
                      (if (and (seq v) (= "" (peek v)))
                        (recur (pop v))
                        v))
            prefix  (if (seq trimmed) [""] [])
            heading-line (str "* " section)
            new-lines (vec (concat trimmed prefix [heading-line ""] block-lines))]
        {:content (str (str/replace (str/join "\n" new-lines) #"\n*$" "") "\n")
         :line    (inc (- (count new-lines) (count block-lines)))})

      (nil? heading-idx)
      nil

      :else
      (let [next-idx (next-top-level-heading-idx lines (inc heading-idx))
            insert-before (or next-idx (count lines))
            ;; Trim trailing blanks in section so splice doesn't push next
            ;; heading further on each insert.
            tail (loop [t insert-before]
                   (if (and (> t (inc heading-idx))
                            (= "" (nth lines (dec t))))
                     (recur (dec t))
                     t))
            insertion (cond-> []
                        (> tail (inc heading-idx)) (conj "")
                        true                       (into block-lines)
                        (some? next-idx)           (conj ""))
            new-lines (vec (concat (subvec lines 0 tail)
                                   insertion
                                   (subvec lines insert-before)))
            heading-line-offset (+ tail (if (= "" (first insertion)) 1 0))]
        {:content (str (str/replace (str/join "\n" new-lines) #"\n*$" "") "\n")
         :line    (inc heading-line-offset)}))))

(defn insert-subtree-into-section
  "Purely splice an already-rendered task subtree below a level-1 section.
  Returns `{:content :line}` or nil when the section does not exist; unlike
  create, restoration never silently creates a destination section."
  [content section subtree]
  (splice-into-section content section subtree false))

;; ── Idempotency scan ───────────────────────────────────────────────

(defn- safe-slurp [path]
  (loader/safe-slurp path))

(defn- walk-tasks-with-imports
  "Return `[{:task, :file}]` for every task reachable from `paths` via
  `#+IMPORT:` (file-level + per-task), de-duped by absolute path."
  [project-root paths]
  (let [visited (volatile! #{})
        out     (volatile! [])]
    (letfn [(walk-file [abs-path]
              (when (and abs-path (not (contains? @visited abs-path)))
                (vswap! visited conj abs-path)
                (when-let [content (safe-slurp abs-path)]
                  (let [effective-content (effective/read-effective-org-content
                                           project-root abs-path content)
                        {:keys [tasks file-imports]}
                        (parser/parse-tasks content {:source-path abs-path
                                                     :effective-source-content effective-content})]
                    (recurse-tasks tasks abs-path)
                    (doseq [imp file-imports]
                      (when-let [abs (links/resolve-link-target
                                       project-root abs-path effective-content imp)]
                        (walk-file abs)))))))
            (recurse-tasks [ts file]
              (doseq [t ts]
                (vswap! out conj {:task t :file file})
                (recurse-tasks (:children t) file)
                (when-let [abs (links/resolve-task-link-target project-root t (:import-path t))]
                  (walk-file abs))))]
      (doseq [p paths] (walk-file p)))
    @out))

(defn- find-duplicate
  "Find the first task whose `:LINKED_ISSUES:` overlaps with `tokens`."
  [collected tokens]
  (when (seq tokens)
    (let [wanted (set tokens)]
      (some
        (fn [{:keys [task file]}]
          (when-let [linked (parser/get-drawer-property task "LINKED_ISSUES")]
            (some
              (fn [tok]
                (when (contains? wanted tok)
                  {:status :duplicate
                   :existing-id (parser/get-task-id task)
                   :existing-file file
                   :conflicting-token tok}))
              (filter seq (str/split linked #"\s+")))))
        collected))))

;; ── Source-line insertion (parent / after) ────────────────────────

(defn- relevel-block [block level]
  (str/replace-first block #"^\*+" (apply str (repeat level "*"))))

(defn- insert-block-before-line [content line-1-indexed block]
  (let [lines (vec (str/split content #"\n" -1))
        idx   (max 0 (dec line-1-indexed))
        block-lines (vec (str/split (str/replace block #"\n+$" "") #"\n"))
        insertion (cond-> []
                    (and (pos? idx) (not= "" (nth lines (dec idx) ""))) (conj "")
                    true (into block-lines)
                    (not= "" (nth lines idx "")) (conj ""))
        next-lines (vec (concat (subvec lines 0 idx) insertion (subvec lines idx)))]
    {:content (str (str/replace (str/join "\n" next-lines) #"\n*$" "") "\n")
     :line (inc (+ idx (if (= "" (first insertion)) 1 0)))}))

;; ── Public entry ───────────────────────────────────────────────────

(defn insert-task-into-file
  "Insert a new task block into `:file` under `:section`. Args mirror
  `build-task-block` plus:

    :file                     - target org file path (cwd-relative or absolute)
    :section                  - level-1 section name
    :allow-create-section?    - default false; when true, missing section
                                is appended
    :also-scan                - additional files scanned for
                                :LINKED_ISSUES: collisions (recursively
                                walks `#+IMPORT:`)
    :project-root             - sandbox root; defaults to cwd

  Returns one of:

    {:status :inserted, :id, :file, :line}
    {:status :duplicate, :existing-id, :existing-file, :conflicting-token}
    {:status :section-not-found, :file, :section}
    {:status :error, :reason, :message}"
  [{:keys [file project-root section summary linked-issues also-scan
           allow-create-section? parent-id after-id]
    :or {project-root (System/getProperty "user.dir")
         also-scan []
         allow-create-section? false}
    :as args}]
  (cond
    (or (nil? summary) (str/blank? summary))
    {:status :error :reason :empty-summary
     :message "`summary` is required and must be non-empty."}

    :else
    (let [target-abs (paths/resolve-project-path
                       project-root project-root file)]
      (cond
        (nil? target-abs)
        {:status :error :reason :path-outside-project
         :message (str "Target file resolves outside project root: " file)}

        :else
        (let [scan-paths-or-error
              (reduce (fn [acc candidate]
                        (if-let [sandboxed (paths/resolve-project-path
                                             project-root project-root candidate)]
                          (conj acc sandboxed)
                          (reduced {:status :error :reason :path-outside-project
                                    :message (str "Scan file resolves outside project root: "
                                                  candidate)})))
                      [target-abs]
                      (or also-scan []))]
          (if (= :error (:status scan-paths-or-error))
            scan-paths-or-error
            (let [scan-paths scan-paths-or-error
                  tokens (filterv #(and % (seq %)) (or linked-issues []))
                  duplicate (when (seq tokens)
                              (let [collected (walk-tasks-with-imports
                                                project-root scan-paths)]
                                (find-duplicate collected tokens)))]
              (if duplicate
                duplicate
                (let [built (build-task-block args)
                      existing (or (safe-slurp target-abs) "")]
                  (if (or parent-id after-id)
                    (if (str/blank? existing)
                      {:status :section-not-found
                       :file target-abs
                       :section section}
                      (let [effective-content (effective/read-effective-org-content
                                               project-root target-abs existing)
                            parsed (parser/parse-tasks
                                    existing
                                    {:source-path target-abs
                                     :effective-source-content effective-content})
                            parent (tree/find-by-id (:tasks parsed) parent-id {:imports? false})
                            after  (tree/find-by-id (:tasks parsed) after-id {:imports? false})]
                        (cond
                          (and parent-id (nil? parent))
                          {:status :unknown-task
                           :reason :parent-not-found
                           :message (str "Parent task not found: " parent-id)}

                          (and after-id (nil? after))
                          {:status :unknown-task
                           :reason :after-not-found
                           :message (str "Anchor task not found: " after-id)}

                          :else
                          (let [anchor (or after parent)
                                level  (if parent
                                         (inc (:level parent))
                                         (:level anchor))
                                block  (relevel-block (:block built) level)
                                spliced (insert-block-before-line
                                         existing (:end-line anchor) block)]
                            (loader/atomic-write target-abs (:content spliced))
                            {:status :inserted
                             :id (:id built)
                             :file target-abs
                             :line (:line spliced)}))))
                    (let [spliced (splice-into-section
                                    existing section (:block built)
                                    allow-create-section?)]
                      (cond
                        (and (not spliced) (not allow-create-section?))
                        {:status :section-not-found
                         :file target-abs
                         :section section}

                        :else
                        (do (loader/atomic-write target-abs (:content spliced))
                            {:status :inserted
                             :id (:id built)
                             :file target-abs
                             :line (:line spliced)})))))))))))))
